Release notes (2021-11-02) (#6854)
keystonejs/keystone: docs/pages/releases/2021-11-01.mdx

```
## Live Reloading <Emoji symbol="🚀" alt="Rocket" />

Keystone now supports live reloading with `keystone-next dev`

You can now update your GraphQL schema, change your hooks and access control, log errors and see how your data returns, then immediately use the playground to test it and iterate.

This is in addition to the current support for live reloading changes to custom views in the Admin UI.

### How it works

When you run `keystone-next dev` now, it will start watching for changes to your config and schema. When changes are detected, Keystone will reload your config and schema and hot-swap the GraphQL endpoint.

To balance performance and predictability, Keystone does not do a complete restart when changes are made. The things to know are:

#### Prisma Schema Changes

The Prisma Client is **not** reloaded as you make changes to the Keystone Schema. If you make changes that affect your Prisma schema, Keystone will exit the process and wait for you to run `keystone-next dev` again.

This is because making database schema changes involves generating database migrations, and can result in data loss if those migrations are automatically run against your current database.

> When adding or removing lists and fields (except virtual fields), we recommend you finish making your changes then start the `dev` process again. This will generate a single migration for all the changes you have made, and interactively prompt you for how to deal with migrations that can't be run safely.

Reviewer: ```suggestion
When adding or removing lists and fields (except virtual fields), we recommend you finish making your changes then start the `dev` process again. This will generate a single migration for all the changes you’ve made, and interactively prompt you to resolve any migrations that can't be run safely.
```

## Removed `resolveFields` <Emoji symbol="🚧" alt="Construction" />

The deprecated `resolveFields` from `context.query` has been removed.

If you were still using it, you should switch to providing `the query option` to `context.query` or use `context.db` if you were providing `false`.

The `context.query` functions will now also throw an error if an empty string is passed to `query` rather than silently returning what the `context.db` functions return, you must select at least one field or omit the `query` option to default to selecting the `id`.

## Internal types renamed <Emoji symbol="✏️" alt="Pencil" />

We have updated `@graphql-ts/schema` to `0.5.0`.

The `__rootVal` properties on `ObjectType`, `InterfaceType` and `UnionType` have been renamed to `__source`, this is intended to be internal but you may have depended on it, if you were using it, rename `__rootVal` to `__source`.

In addition the `fields` property on `InterfaceType` has been renamed to `__fields` and it will no longer exist at runtime like the other types.

## GraphQL Schema Reorder <Emoji symbol="🛗" alt="Elevator" />

> The way the `endSession` field on the `Mutation` type and the `keystone` field on the `Query` type have been adjusted. This may result in re-ordering in your generated `schema.graphql` file. The `sessionSchema` export of `@keystone-next/keystone/session` has also been removed.

Reviewer: "The way the `endSession` field on the `Mutation` type and the `keystone` field on the `Query` type have been adjusted."

^ I find this hard to understand. 

Is it the `order` of those things?

Author: I tweaked this a little.

## Miscellaneous Fixes <Emoji symbol="⚙️" alt="Gear" />

- `context.session` no longer has a value if the session is invalid.

> - `text`, `integer`, `float` and `decimal` on the item view when using `ui.itemView.fieldMode: 'read'` now render correctly.

Reviewer: ```suggestion
- `text`, `integer`, `float` and `decimal` on the item view now render correctly when using `ui.itemView.fieldMode: 'read'`.
```

- Admin UI home page now respects `ui.hideCreate` and won't show a plus button when create is disabled.

> - Read only arrays are now accepted where previously mutable arrays were required. This means that if you use `as const` when writing an array and then pass it to various APIs in Keystone, that will now work.

Reviewer: ```suggestion
- Read-only arrays are now accepted where previously mutable arrays were required. Using `as const` when writing an array and then passing it to various APIs in Keystone will now work.
```

In addition the `fields` property on `InterfaceType` has been renamed to `__fields` and it will no longer exist at runtime like the other types.

## GraphQL Schema Reorder <Emoji symbol="🛗" alt="Elevator" />

We've made internal changes to the `endSession` field on the `Mutation` type and the `keystone` field on the `Query` type. This may result in re-ordering in your generated `schema.graphql` file.

## Miscellaneous Fixes <Emoji symbol="⚙️" alt="Gear" />

- The `sessionSchema` export of `@keystone-next/keystone/session` has been removed.
- `context.session` no longer has a value if the session is invalid.
- `text`, `integer`, `float` and `decimal` on the item view now render correctly when using `ui.itemView.fieldMode: 'read'`.
- Admin UI home page now respects `ui.hideCreate` and won't show a plus button when create is disabled.
- Read-only arrays are now accepted where previously mutable arrays were required. Using `as const` when writing an array and then passing it to various APIs in Keystone will now work.
- Fixed bug in `LinkToRelatedItems` button for double sided relationships.

> - The `src` field on the output of `image` and `file` fields has been renamed to `url`, subsequently the `getSrc` function on `ImagesContext` and `FilesContext` has been renamed to `getUrl`.

Reviewer 2: Should this be a breaking change?
